packages/block-editor/src/components/block-list/insertion-point.js

```diff
@@ -6,93 +6,102 @@ import classnames from 'classnames';
-		this.onFocusInserter = this.onFocusInserter.bind( this );
-	}
+function BlockInsertionPoint( { rootClientId, clientId } ) {
+	const [ isInserterFocused, setIsInserterFocused ] = useState( false );
 
-	onFocusInserter( event ) {
+	const onFocusInserter = useCallback( ( event ) => {
 		// Stop propagation of the focus event to avoid selecting the current
 		// block while inserting a new block, as it is not relevant to sibling
 		// insertion and conflicts with contextual toolbar placement.
 		event.stopPropagation();
 
-		this.setState( {
-			isInserterFocused: true,
-		} );
-	}
+		setIsInserterFocused( true );
+	}, [] );
 
-	onBlurInserter() {
-		this.setState( {
-			isInserterFocused: false,
-		} );
-	}
+	const onBlurInserter = useCallback( () => setIsInserterFocused( false ), [] );
 
-	render() {
-		const { isInserterFocused } = this.state;
+	const { showInsertionPoint } = useSelect( ( select ) => {
 		const {
-			showInsertionPoint,
-			rootClientId,
-			clientId,
-		} = this.props;
+			getBlockIndex,
+			getBlockInsertionPoint,
+			isBlockInsertionPointVisible,
+		} = select( 'core/block-editor' );
+		const blockIndex = getBlockIndex( clientId, rootClientId );
+		const insertionPoint = getBlockInsertionPoint();
+		return {
+			showInsertionPoint: (
+				isBlockInsertionPointVisible() &&
+				insertionPoint.index === blockIndex &&
+				insertionPoint.rootClientId === rootClientId
+			),
+		};
+	}, [ clientId, rootClientId ] );
+
+	const { selectBlock, selectPreviousBlock } = useDispatch( 'core/block-editor' );
+
+	const selectAdjacentBlock = useCallback( ( event ) => {
+		const { currentTarget } = event;
+		const targetRect = currentTarget.getBoundingClientRect();
+		const yPercent = ( event.clientY - targetRect.top ) / targetRect.height;
+
+		selectBlock( null );
 
-		return (
-			<div className="editor-block-list__insertion-point block-editor-block-list__insertion-point">
-				{ showInsertionPoint && (
-					<div className="editor-block-list__insertion-point-indicator block-editor-block-list__insertion-point-indicator" />
-				) }
-				<div
-					onFocus={ this.onFocusInserter }
-					onBlur={ this.onBlurInserter }
-					// While ideally it would be enough to capture the
-					// bubbling focus event from the Inserter, due to the
-					// characteristics of click focusing of `button`s in
-					// Firefox and Safari, it is not reliable.
-					//
-					// See: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
-					tabIndex={ -1 }
-					className={
-						classnames( 'editor-block-list__insertion-point-inserter block-editor-block-list__insertion-point-inserter', {
-							'is-visible': isInserterFocused,
-						} )
-					}
-				>
-					<Inserter
-						rootClientId={ rootClientId }
-						clientId={ clientId }
-					/>
-				</div>
+		const caretRect = new window.DOMRect( event.clientX, targetRect.top, 0, 28 );
+		if ( yPercent > 0.5 ) {
+			selectBlock( clientId, { caretRect } );
```

Comment:
Why not call `placeCaretAtVerticalEdge` directly and avoid the strange `initialPosition` state?

Reply:
> Why not call `placeCaretAtVerticalEdge` directly and avoid the strange `initialPosition` state?

It could be done, but the difficulty is in finding the input, which likely would require copying one or the other of WritingFlow or BlockListBlock's respective behaviors for finding a tabbable at the extent of the block [[1]](https://github.com/WordPress/gutenberg/blob/becdb52c0d1dbdd72b296689b32f809ff208188e/packages/block-editor/src/components/writing-flow/index.js#L112-L174) [[2]](https://github.com/WordPress/gutenberg/blob/becdb52c0d1dbdd72b296689b32f809ff208188e/packages/block-editor/src/components/block-list/block.js#L209-L236).

Maybe there's another option where we "wait" for the focus to take place from the block selection, and then call `placeCaretAtVerticalEdge` from the `document.activeElement`.